### PR TITLE
ci: fix the minikube version check failure

### PR DESCRIPTION
**Pull request by pkalever**
## Describe what this PR does ##

 Problem:                                                                        
 ------------                                                                         
Fix the minikube version check failure
                                                                                 
 $ minikube version                                                              
 minikube version: v1.12.2                                                       
 commit: be7c19d391302656d27f1f213657d925c4e1cfc2-dirty                          
                                                                                 
 $ ./scripts/minikube.sh up                                                      
 installed minikube version v1.12.2 is not matching requested version latest     
                                                                                 
 Here v1.12.2 is the latest version of minikube, but the script simply bails out.

### CODE REVIEW

File: scripts/minikube.sh [hunk 1]
```diff
@@ -113,7 +113,7 @@ function minikube_supports_psp() {
 
 # configure minikube
 MINIKUBE_ARCH=${MINIKUBE_ARCH:-"amd64"}
-MINIKUBE_VERSION=${MINIKUBE_VERSION:-"latest"}
+MINIKUBE_VERSION=${MINIKUBE_VERSION:-"v1.12.2"}
```
Comment (Madhu-1) [inline scripts/minikube.sh:116]: This is not always true. what if we get a new minikube release.

File: scripts/minikube.sh [hunk 2]
```diff
@@ -99,6 +98,19 @@ function minikube_losetup() {
     ${minikube} ssh 'sudo sh -c "rm -f /sbin/losetup && cp ~docker/losetup /sbin"'
 }
 
+function minikube_supports_psp() {
```
Comment (Madhu-1) [inline scripts/minikube.sh:101]: please update the commit message why the reordering is required

File: scripts/minikube.sh [hunk 3]
```diff
@@ -77,7 +76,7 @@ function enable_psp() {
     echo "prepare minikube to support pod security policies"
     mkdir -p "$HOME"/.minikube/files/etc/kubernetes/addons
     DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
-	  cp "$DIR"/psp.yaml "$HOME"/.minikube/files/etc/kubernetes/addons/psp.yaml
+    cp "$DIR"/psp.yaml "$HOME"/.minikube/files/etc/kubernetes/addons/psp.yaml
```
Comment (Madhu-1) [inline scripts/minikube.sh:79]: This can be a separate commit this should not be part of this commit

File: scripts/minikube.sh [hunk 4]
```diff
@@ -99,9 +98,22 @@ function minikube_losetup() {
...
+    if [[ "${MINIKUBE_MAJOR}" -ge 1 ]] && [[ "${MINIKUBE_MINOR}" -ge 11 ]] && [[ "${MINIKUBE_PATCH}" -ge 1 ]] || [[ "${MINIKUBE_MAJOR}" -ge 1 ]] && [[ "${MINIKUBE_MINOR}" -ge 12 ]]; then
+        return 1
+    fi
+    return 0
+}
+
 # configure minikube
 MINIKUBE_ARCH=${MINIKUBE_ARCH:-"amd64"}
-MINIKUBE_VERSION=${MINIKUBE_VERSION:-"latest"}
+MINIKUBE_VERSION=${MINIKUBE_VERSION:-"v1.12.2"}
```
Comment (nixpanic) [inline scripts/minikube.sh:116]: Fixed versions should only be in the `build.env` file. When trying to install minikube with `latest`, I think it should not check the currently installed minikube version, but just install it.

There might be a minikube command to check if there is an update available. That could be used in addition to the above.

File: scripts/minikube.sh (hunk 1 above)
Comment (pkalever) [reply scripts/minikube.sh:116]: We should do the same thing what we do for KUBE_VERSION variable in this script :-)

or maybe we should not care for which version of minikube is in-use, like get rid of below:
`if [[ "${version}" != "${MINIKUBE_VERSION}" ]]; then`

File: scripts/minikube.sh (hunk 4 above)
Comment (pkalever) [reply scripts/minikube.sh:116]: +1 @nixpanic

File: scripts/minikube.sh (hunk 3 above)
Comment (pkalever) [reply scripts/minikube.sh:79]: Sure Madhu, will update.

File: scripts/minikube.sh (hunk 2 above)
Comment (pkalever) [reply scripts/minikube.sh:101]: Yep. Until then:

Generally, we define all the functions at the top and start the runnable commands (entry points to script start).
Only this function is odd in the script unlike the rest of them taking away the readability.

**Comment by pkalever** [comment]
> we don't have minikube component for the commit please go through https://github.com/ceph/ceph-csi/blob/master/docs/development-guide.md#code-contribution-workflow to know more about it

Request not to take the commit msgs as direct indication of the component. Instead, they just refer to some context.
Like it could be a directory name, file name, or the component itself.

One close example:
```
commit 58c2a3f15fdd176dc3165082c49f5b7ef5fe7970                          
Author:     Niels de Vos <ndevos@redhat.com>                             
AuthorDate: Sat Apr 18 16:47:56 2020 +0200                               
Commit:     mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>
CommitDate: Tue Apr 21 05:26:32 2020 +0000                               

    minikube: fix ShellCheck errors

    Signed-off-by: Niels de Vos <ndevos@redhat.com>                      
```

I think having `minikube:` doesn't hurt as it is the one which we are playing around in this change, but if you insist, I can change it for you :-)

Thanks!

**Comment by Madhu-1** [comment]
> > we don't have minikube component for the commit please go through https://github.com/ceph/ceph-csi/blob/master/docs/development-guide.md#code-contribution-workflow to know more about it
> 
> Request not to take the commit msgs as direct indication of the component. Instead, they just refer to some context.
> Like it could be a directory name, file name, or the component itself.
> 
> One close example:
> 
> ```
> commit 58c2a3f15fdd176dc3165082c49f5b7ef5fe7970                          
> Author:     Niels de Vos <ndevos@redhat.com>                             
> AuthorDate: Sat Apr 18 16:47:56 2020 +0200                               
> Commit:     mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>
> CommitDate: Tue Apr 21 05:26:32 2020 +0000                               
> 
>     minikube: fix ShellCheck errors
> 
>     Signed-off-by: Niels de Vos <ndevos@redhat.com>                      
> ```
> 
> I think having `minikube:` doesn't hurt as it is the one which we are playing around in this change, but if you insist, I can change it for you :-)
> 
> Thanks!

For the above example was before we made some standards for commit messages. we do request contributors to go through https://github.com/ceph/ceph-csi/blob/master/docs/development-guide.md#commit-messages and understand the current standard we follow in ceph-csi repo.

**Comment by nixpanic** [comment]
> I think having `minikube:` doesn't hurt as it is the one which we are playing around in this change, but if you insist, I can change it for you :-)

The valid components are now listed in `.commitlintrc.yml`. Instead of `minikube:`, `ci:` would be more appropriate.

File: scripts/minikube.sh [hunk 5]
```diff
@@ -54,10 +54,6 @@ function install_minikube() {
         local mk_version version
         read -ra mk_version <<<"$(${minikube} version)"
         version=${mk_version[2]}
```
Comment (nixpanic) [inline scripts/minikube.sh:56]: If the requested version is `latest`, this check can be skipped. In other cases, it should be kept.

**Review by nixpanic — CHANGES_REQUESTED** [review]

File: scripts/minikube.sh (hunk 5 above)
Comment (pkalever) [reply scripts/minikube.sh:56]: @nixpanic have you looked at the next new commit, where the commit tries to update the minikube if the latest version exists.

Thanks!

File: scripts/minikube.sh [hunk 6]
```diff
@@ -55,7 +55,17 @@ function install_minikube() {
...
         echo "minikube already installed with ${version}"
-        return
+        mku_version=$(minikube update-check 2> /dev/null || ret=${?})
```
Comment (nixpanic) [inline scripts/minikube.sh:58]: you need to execute `${minikube}` like done in other places.

`ret` is set in a subshell (started with `$(...)`), I would be surprised if the value of `ret` is stored in `ret` after the subshell exits. Please check.

File: scripts/minikube.sh (hunk 5 above)
Comment (nixpanic) [reply scripts/minikube.sh:56]: I somehow missed that, thanks!

File: scripts/minikube.sh [hunk 7]
```diff
@@ -55,7 +55,17 @@ function install_minikube() {
...
+        latest_version=$(echo ${mku_version} | cut -d' ' -f4)
+        if [[ ${ret} -ne 0 ]] || [[ -z ${latest_version} ]]; then
```
Comment (nixpanic) [inline scripts/minikube.sh:60]: Place variables in `"` quotes please. I guess `shellcheck` will complain about that too.

**Review by nixpanic — CHANGES_REQUESTED** [review]

File: scripts/minikube.sh (hunk 6 above)
Comment (pkalever) [reply scripts/minikube.sh:58]: > you need to execute `${minikube}` like done in other places.

Sure.

> 
> `ret` is set in a subshell (started with `$(...)`), I would be surprised if the value of `ret` is stored in `ret` after the subshell exits. Please check.

Yes, it will. The trick was played here because simply using:
mku_version=$(minikube update-check)
will exit the whole script if the `minikube update-check` command fails, like in case the network is not available etc..

In the case of failure the above trick will 1. help not exit the script 2. collect the exit code of the command too

Thanks!

File: scripts/minikube.sh (hunk 7 above)
Comment (pkalever) [reply scripts/minikube.sh:60]: sure will take care of this.

File: scripts/minikube.sh [hunk 8]
```diff
@@ -51,11 +51,21 @@ function detect_minikube() {
 # install minikube
 function install_minikube() {
     if type "${minikube}" >/dev/null 2>&1; then
-        local mk_version version
+        local mk_version version mku_version latest_version ret
         read -ra mk_version <<<"$(${minikube} version)"
         version=${mk_version[2]}
         echo "minikube already installed with ${version}"
-        return
+        mku_version=$(${minikube} update-check 2> /dev/null || ret=${?})
```
Comment (nixpanic) [inline scripts/minikube.sh:58]: shellcheck (part of `make-containerized-test`) agrees that changing `ret` in a sub-shell is not working as you intend.

**Review by nixpanic — CHANGES_REQUESTED** [review]

File: scripts/minikube.sh (hunk 8 above)
Comment (pkalever) [reply scripts/minikube.sh]: I had tested it enabling `set -x` with various conditions and it worked for me and logically it looks fine too :-)
Any have, now that it bothers shellcheck, I had made some changes and updated the PR. Please take a look.

Thanks @nixpanic

File: scripts/minikube.sh [hunk 9]
```diff
@@ -50,16 +50,28 @@ function detect_minikube() {
 
 # install minikube
 function install_minikube() {
+     local mku_version latest_version
+     mku_version=$(${minikube} update-check 2> /dev/null || true)
+     latest_version=$(echo ${mku_version} | cut -d' ' -f4)
```
Comment (nixpanic) [inline scripts/minikube.sh:55]: I have an old version of minikube, that outputs:
```
$ minikube update-check
CurrentVersion: v0.27.0
LatestVersion: v1.12.3
```
Better set `latest_version` to `awk '/LatestVersion {print $NF}' <<< "${mku_version}"` so that the order of the lines, or additions do not influence the result.

This should also probably be inside the `if type ${minikube}` statement in case minikube is not installed yet.

Comment (nixpanic) [reply scripts/minikube.sh:55]: Indention looks a little off too, compared to the original lines below.

File: scripts/minikube.sh [hunk 10]
```diff
@@ -50,16 +50,28 @@ function detect_minikube() {
...
+     fi
     if type "${minikube}" >/dev/null 2>&1; then
         local mk_version version
         read -ra mk_version <<<"$(${minikube} version)"
         version=${mk_version[2]}
         echo "minikube already installed with ${version}"
-        return
+        MINIKUBE_VERSION=${version}
+        if [[ "${version}" == "${latest_version}" ]]; then
+          echo "minikube is already the latest version"
```
Comment (nixpanic) [inline scripts/minikube.sh:67]: weird indention, make it 4 spaces please.

**Review by nixpanic — CHANGES_REQUESTED** [review]

File: scripts/minikube.sh (hunk 9 above)
Comment (pkalever) [reply scripts/minikube.sh:55]: > I have an old version of minikube, that outputs:
> 
> ```
> $ minikube update-check
> CurrentVersion: v0.27.0
> LatestVersion: v1.12.3
> ```
> 
> Better set `latest_version` to `awk '/LatestVersion {print $NF}' <<< "${mku_version}"` so that the order of the lines, or additions do not influence the result.

This won't work Niels, as storing the output of the command will already convert the above multiline output to single line.

✨ mku_version=$(minikube update-check 2> /dev/null || true)
✨ echo $mku_version
CurrentVersion: v1.12.3 LatestVersion: v1.12.3

But I get your point, will defend on it.

> 
> This should also probably be inside the `if type ${minikube}` statement in case minikube is not installed yet.

I'm trying to capture the latest_version which will be used after the if block, and absence of minkube command will be taken care, as it will result in empty string ${mku_version}

Comment (pkalever) [reply scripts/minikube.sh:55]: > Indention looks a little off too, compared to the original lines below.

Hmm, tweaked my vimrc, thanks!

File: scripts/minikube.sh (hunk 10 above)
Comment (pkalever) [reply scripts/minikube.sh]: tweaked my vimrc now, thanks!

File: scripts/minikube.sh [hunk 11]
```diff
@@ -50,16 +50,28 @@ function detect_minikube() {
...
+        # skip: update-check failed for some reason, lets continue with what we have
+        latest_version=${MINIKUBE_VERSION}
+    fi
     if type "${minikube}" >/dev/null 2>&1; then
         local mk_version version
         read -ra mk_version <<<"$(${minikube} version)"
         version=${mk_version[2]}
         echo "minikube already installed with ${version}"
-        return
+        MINIKUBE_VERSION=${version}
```
Comment (nixpanic) [inline scripts/minikube.sh:65]: Is it needed to change the `MINIKUBE_VERSION` in this function? If it contains `latest`, that should be fine, I think?

Comment (pkalever) [reply scripts/minikube.sh:65]: Just feeding it with the real version number for future use `within the script`.
Here it `might` be a lower version, and at the end, once the update is done then we assign with the latest.

@nixpanic do you see any issue with this?

Comment (pkalever) [reply scripts/minikube.sh:65]: This way MINIKUBE_VERSION is always set to the currently installed version rather than faking as `lastest` always.

Comment (nixpanic) [reply scripts/minikube.sh:65]: Because `MINIKUBE_VERSION` gets passed into the script through an environment variable, changing its value is ugly. If there is a need for using the actual version later on, it is cleaner to add a new (lowercase) variable for that.

Comment (pkalever) [reply scripts/minikube.sh]: @nixpanic ok, updated addressing the review comments. Thanks!

**Comment by pkalever** [comment]
@Madhu-1 @nixpanic FYI the CI is failing with:
```
error: unable to read URL "https://raw.githubusercontent.com/kubernetes-csi/external-snapshotter/master/config/crd/snapshot.storage.k8s.io_volumesnapshotclasses.yaml", server reported 404 Not Found, status code=404
```

**Comment by Madhu-1** [comment]
> @Madhu-1 @nixpanic FYI the CI is failing with:
> 
> ```
> error: unable to read URL "https://raw.githubusercontent.com/kubernetes-csi/external-snapshotter/master/config/crd/snapshot.storage.k8s.io_volumesnapshotclasses.yaml", server reported 404 Not Found, status code=404
> ```

rebase with the latest master.

**Review by nixpanic — APPROVED** [review]

**Review by Madhu-1 — APPROVED** [review]
